feat: add `logger.hasErrorLogged(error)` method

aleclarson opened the pull request:
<!-- Thank you for contributing! -->

### Description

This makes it easier for frameworks which wrap Vite (eg: `vite-plugin-ssr`) to avoid logging an error that was already logged by Vite.

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

aleclarson committed:
feat: set `loggedByVite` on errors logged by Vite

yyx990803 commented:
Let's pass the error to `logger.error` as part of the options and do the marking inside `logger.error`

aleclarson committed:
feat: add `logger.hasLogged(error)` method

..and require an `error` option to be defined when `logger.error` is called.

BREAKING CHANGE

aleclarson commented:
I've added the `logger.hasLogged` method, and the `error` option **must be defined** by callers of `logger.error`. They can pass `error: null` if no associated error exists. This ensures the option isn't forgotten about.

In Vite itself, most logged errors are **not** rethrown, so marking them is pointless, but I'm passing the `Error` object whenever possible as a principle.

aleclarson committed:
chore: update logger.error callsites

antfu commented on packages/vite/src/node/logger.ts:
```diff
@@ -51,13 +56,18 @@ export function createLogger(
 ): Logger {
   const { prefix = '[vite]', allowClearScreen = true } = options
 
+  const loggedErrors = new WeakMap<Error | RollupError, boolean>()
```
Could we simply use WeakSet?

aleclarson replied:
Oh, good point. Yeah let's do that

antfu committed:
chore: update

antfu committed:
chore: merge

antfu committed:
Merge branch 'main' into feat/loggedByVite

patak-js commented:
I see that there is a comment about a breaking change, @antfu is this safe to merge in the 2.5 beta? We'll do a beta-1 today so we could still include this one

antfu committed:
chore: refactoring

antfu committed:
chore: update types

antfu reviewed (dismissed):
I have changed the `options` argument to optional to avoid breaking changes.

antfu committed:
chore: export types
